rcl_action/src/rcl_action/graph.c

```diff
@@ -0,0 +1,202 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <assert.h>
+#include <string.h>
+
+#include "rcl/error_handling.h"
+#include "rcl/graph.h"
+#include "rcl/node.h"
+#include "rcutils/strdup.h"
+
+#include "rcl_action/graph.h"
+
+static
+rcl_ret_t
+_filter_action_names(
+  rcl_names_and_types_t * topic_names_and_types,
+  rcl_allocator_t * allocator,
+  rcl_names_and_types_t * action_names_and_types)
+{
+  assert(topic_names_and_types);
+  assert(allocator);
+  assert(action_names_and_types);
+
+  // Assumption: actions provide a topic name with the suffix "/_action/feedback"
+  // and it has type with the suffix "_FeedbackMessage"
+  const char * action_name_suffix = "/_action/feedback";
+  const char * action_type_suffix = "_FeedbackMessage";
+
+  rcl_ret_t ret;
+  const size_t num_names = topic_names_and_types->names.size;
+  char ** names = topic_names_and_types->names.data;
+
+  // Count number of actions to determine how much memory to allocate
+  size_t num_actions = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * suffix_index = strstr(names[i], action_name_suffix);
+    if (suffix_index) {
+      num_actions++;
+    }
+  }
+
+  if (0u == num_actions) {
+    return RCL_RET_OK;
+  }
+
+  ret = rcl_names_and_types_init(action_names_and_types, num_actions, allocator);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  ret = RCL_RET_OK;
+
+  // Prune names/types that are not actions (ie. do not contain the suffix)
+  const size_t suffix_len = strlen(action_name_suffix);
+  size_t j = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * suffix_index = strstr(names[i], action_name_suffix);
+    if (suffix_index) {
+      const size_t action_name_len = strlen(names[i]) - suffix_len;
+      char * action_name = rcutils_strndup(names[i], action_name_len, *allocator);
+      if (!action_name) {
+        RCL_SET_ERROR_MSG("Failed to allocate memory for action name");
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      action_names_and_types->names.data[j] = action_name;
+
+      // Allocate storage for type list
+      rcutils_ret_t rcutils_ret = rcutils_string_array_init(
+        &action_names_and_types->types[j],
+        topic_names_and_types->types[i].size,
+        allocator);
+      if (RCUTILS_RET_OK != rcutils_ret) {
+        RCL_SET_ERROR_MSG(rcutils_get_error_string().str);
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      // Populate types list
+      for (size_t k = 0u; k < topic_names_and_types->types[i].size; ++k) {
+        char * type_name = topic_names_and_types->types[i].data[k];
+        size_t action_type_len = strlen(type_name);
+        // Trim type name suffix, if provided
+        if (action_type_suffix) {
```

Comment:
This is always true, unless this variable is shadowed and I missed it

```
 const char * action_type_suffix = "_FeedbackMessage";
```

Reply:
Yep. Removed 2923012